### PR TITLE
[MXNET-535] Fix bugs in LR Schedulers and add warmup

### PR DESCRIPTION
**Pull request by rahul003**
## Description ##
- Adds warmup to all LR schedulers. Of two modes, linear increase and constant warmup
- Also fixes inconsistencies/bugs where base_lr is not taken by MultiFactorScheduler and FactorScheduler.
- Added tests for LR schedulers

### Essentials ###
Please feel free to remove inapplicable items for your PR.
- [x] The PR title starts with [MXNET-$JIRA_ID], where $JIRA_ID refers to the relevant [JIRA issue](https://issues.apache.org/jira/projects/MXNET/issues) created (except PRs with tiny changes)
- [x] Changes are complete (i.e. I finished coding on this PR)
- [x] All changes have test coverage:
- Unit tests are added for small changes to verify correctness (e.g. adding a new operator)
- Nightly tests are added for complicated/long-running ones (e.g. changing distributed kvstore)
- Build tests will be added for build configuration changes (e.g. adding a new build option with NCCL)
- [x] Code is well-documented: 
- For user-facing API changes, API doc string has been updated. 
- For new C++ functions in header files, their functionalities and arguments are documented. 
- For new examples, README.md is added to explain the what the example does, the source of the dataset, expected performance on test set and reference to the original paper if applicable
- Check the API doc at http://mxnet-ci-doc.s3-accelerate.dualstack.amazonaws.com/PR-$PR_ID/$BUILD_ID/index.html
- [x] To the my best knowledge, examples are either not affected by this change, or have been fixed to be compatible with this change

### CODE REVIEW

File: python/mxnet/lr_scheduler.py [hunk 1]
```diff
@@ -153,18 +153,57 @@ class PolyScheduler(LRScheduler):
...
+        super(WarmupScheduler, self).__init__()
+        self.lr_begin = lr_begin
+        self.scheduler = scheduler
+        self.lr_final = self.scheduler.base_lr
+        if self.lr_begin > self.lr_final:
+            raise ValueError("Final lr has to be higher than beginning lr")
+        if warmup_steps <= 0:
+            raise ValueError("Warmup steps has to be positive")
+        self.warmup_steps = warmup_steps
+        self.lrs_updates = {}
```
Comment (piiswrong) [inline python/mxnet/lr_scheduler.py]: what's the point of this cache? Looks like it will always miss

File: python/mxnet/lr_scheduler.py [hunk 2]
```diff
@@ -153,18 +153,57 @@ class PolyScheduler(LRScheduler):
...
+        self.lr_difference = self.lr_final - self.lr_begin
+
+    def __call__(self, num_update):
+        if num_update not in self.lrs_updates:
+            if num_update < self.warmup_steps:
+                increase = self.lr_difference * float(num_update)/float(self.warmup_steps)
+                self.lrs_updates[num_update] = self.lr_begin + increase
+            else:
+                if isinstance(self.scheduler, PolyScheduler):
+                    self.lrs_updates[num_update] = self.scheduler(num_update - self.warmup_steps)
```
Comment (piiswrong) [inline python/mxnet/lr_scheduler.py]: Why special case for PolyScheduler?
Is `num_update - self.warmup_steps` standard? Does Tf or Pytorch do it this way?
Why not num_update directly?

Comment (hetong007) [reply python/mxnet/lr_scheduler.py]: PolyScheduler, and CosineScheduler(not implemented here) reduce lr from a "starting lr" to an "ending lr" **smoothly**, for example from 0.1 to 0.

With warmup, we first increase the lr from a small value (e.g. 0) to the starting lr, then apply the main scheduler. Assuming we have warmup for the first 5 epochs, and the total training epochs is 90, then the effective number of epochs for the poly scheduler is 85.

As for piecewise-constant/factor scheduler, it decays the lr only at certain points, on which the warmup stage has no effect.

File: python/mxnet/lr_scheduler.py (hunk 1 above)
Comment (rahul003) [reply python/mxnet/lr_scheduler.py]: We would have for each batch, number of calls to __call__ equal to the number of learnable parameter arrays.

File: python/mxnet/lr_scheduler.py [hunk 3]
```diff
@@ -153,18 +153,57 @@ class PolyScheduler(LRScheduler):
 
     """
 
-    def __init__(self, max_update, base_lr=0.01, pwr=2):
```
[the remark this answers is not in the document]
Comment (rahul003) [reply python/mxnet/lr_scheduler.py]: Now base_lr is going in to super through kwargs. Would that still break the API?

File: python/mxnet/lr_scheduler.py (hunk 2 above)
Comment (rahul003) [reply python/mxnet/lr_scheduler.py]: Yeah for the above reason, but I'm updating the code to remove this special case, and pass a wamup_steps param to such schedulers so we can handle it cleanly

**Comment by rahul003** [comment]
Improved how warmup is handled, please review

**Comment by rahul003** [comment]
@piiswrong Could you please review? This has a fix for an important bug where the MultiFactorScheduler didn't take a base_lr previously. This meant that the example/image_classification/ scripts didn't use the given LR correctly. It would drop from x to 0.001 and 0.0001 regardless of the LR given.

**Comment by rahul003** [comment]
@piiswrong @szha @eric-haibin-lin Please review

**Comment by rahul003** [comment]
@szha @eric-haibin-lin please review

File: python/mxnet/lr_scheduler.py [hunk 4]
```diff
@@ -29,8 +30,31 @@ class LRScheduler(object):
     base_lr : float, optional
         The initial learning rate.
     """
-    def __init__(self, base_lr=0.01):
+    def __init__(self, base_lr=0.01, warmup_steps=0, warmup_begin_lr=0, warmup_mode='linear'):
```
Comment (eric-haibin-lin) [inline python/mxnet/lr_scheduler.py]: Do you mind adding documentation for warmup_begin_lr?

File: python/mxnet/lr_scheduler.py [hunk 5]
```diff
@@ -138,33 +171,73 @@ def __call__(self, num_update):
...
        base_lr:    base learning rate
-       pwr:   power of the decay term as a funtion of the current number of updates.
-
+       final_lr:   final learning rate after all steps
+       warmup_steps: number of warmup steps used before this scheduler starts decay
     """
 
-    def __init__(self, max_update, base_lr=0.01, pwr=2):
-        super(PolyScheduler, self).__init__(base_lr)
+    def __init__(self, max_update, base_lr=0.01, final_lr=0,
```
Comment (piiswrong) [inline python/mxnet/lr_scheduler.py]: why did you remove pwr? This is API breakage

Comment (rahul003) [reply python/mxnet/lr_scheduler.py]: I've not removed it. Git is getting confused. It thinks I've changed PolyScheduler to CosineScheduler when in fact I've modified PolyScheduler and added a new CosineScheduler.

**Comment by rahul003** [comment]
Hopefully this will give committers confidence to merge

## Interfaces
### CosineScheduler
```
def __init__(self, max_update, base_lr=0.01, final_lr=0,
                 warmup_steps=0, warmup_begin_lr=0, warmup_mode='linear'):
```

### PolyScheduler
#### This PR
```
def __init__(self, max_update, base_lr=0.01, pwr=2, final_lr=0,
                 warmup_steps=0, warmup_begin_lr=0, warmup_mode='linear'):
```
#### Earlier
```
def __init__(self, max_update, base_lr=0.01, pwr=2):
```

### MultiFactorScheduler
#### This PR
```
def __init__(self, step, factor=1, base_lr=0.01, warmup_steps=0, warmup_begin_lr=0,
                 warmup_mode='linear'):
```
#### Earlier
```
def __init__(self, step, factor=1):
```

### FactorScheduler
#### This PR
```
def __init__(self, step, factor=1, stop_factor_lr=1e-8, base_lr=0.01,
                 warmup_steps=0, warmup_begin_lr=0, warmup_mode='linear'):
```
#### Earlier
```
def __init__(self, step, factor=1, stop_factor_lr=1e-8):
```

### Scheduler
#### This PR
```
def __init__(self, base_lr=0.01,
                 warmup_steps=0, warmup_begin_lr=0, warmup_mode='linear'):
```
#### Earlier
```
def __init__(self, base_lr=0.01):
```


### Plots of LR decay from unit tests

File: python/mxnet/lr_scheduler.py (hunk 4 above)
Comment (rahul003) [reply python/mxnet/lr_scheduler.py]: There was some for the inherited classes, but not for this base abstract class. Anyway, now added for all. Please check.

File: python/mxnet/lr_scheduler.py (hunk 5 above)
Comment (rahul003) [reply python/mxnet/lr_scheduler.py]: Please refer https://github.com/apache/incubator-mxnet/pull/11234#issuecomment-413998969
